File: src/img_client.cpp
#include<ros/ros.h>
#include<actionlib/client/simple_action_client.h>   // 客户端需要包含的头文件
#include<image_process/ProcImgAction.h> // 自定义Action所要包含的头文件
#include<opencv2/opencv.hpp>    // OpenCV基础包要包含的头文件

// 利用typedef提前定义好类型
typedef actionlib::SimpleActionClient <image_process::ProcImgAction> Client;

// 自定义函数，用于将Message传输的vector<uint8_t>类型数据转换成OpenCV的Mat类型
cv::Mat cvtArray2Img(std::vector<uint8_t> array, int img_height, int img_width) {
    cv::Mat restore_img(img_height, img_width, CV_8UC3);
    for (int i = 0; i < img_height; i++) {
        for (int j = 0; j < img_width; j++) {
            int index = i * img_width + j;
            uint8_t b = array[index];
            uint8_t g = array[index + img_height * img_width];
            uint8_t r = array[index + img_height * img_width * 2];
            restore_img.at<cv::Vec3b>(i, j) = cv::Vec3b(b, g, r);
        }
    }
    return restore_img;
}


// 当Action完成后的回调函数
void doneCb(const actionlib::SimpleClientGoalState &state,
            const image_process::ProcImgResultConstPtr &result) {
    // 新建一个Mat并将传输的数据转换成OpenCV能识别的Mat类型
    cv::Mat processed_img;
    processed_img = cvtArray2Img(result->processed_img, result->img_height, result->img_width);

    // 如果原图太大，将它缩小，这样利于展示
    if (result->img_width > 1200) {
        float ratio = 1200.0 / result->img_width;
        cv::resize(processed_img, processed_img, cv::Size(0, 0), ratio, ratio);
    }

    ROS_INFO("Finished processing image with %d x %d pixels...", result->img_height, result->img_width);

    // 展示图片
    cv::imshow("processed_img", processed_img);
    cv::waitKey(0);

    // 关闭ROS
    ros::shutdown();
}

// 当Action启动后的回调函数
void activeCb() {
    ROS_INFO("Contected to server,start processing image...");
}

// 当收到Feedback后的回调函数
void feedbackCb(const image_process::ProcImgFeedbackConstPtr &feedback) {
    ROS_INFO("percent complete:%.2f%%", feedback->complete_percent);
}

int main(int argc, char *argv[]) {
    // 用户输入图片路径
    std::string img_path;
    std::cout << "Input image path:\n";
    std::cin >> img_path;

    // 初始化并指定节点名称
    ros::init(argc, argv, "img_client");
    // 创建客户端并指定要连接到的服务端名称
    Client client("img_process", true);

    // 等待服务端
    ROS_INFO("Waiting for action server to start.");
    client.waitForServer();
    ROS_INFO("Action server started, sending goal.");

    // 创建Action的goal，传入的是图片路径字符串
    image_process::ProcImgGoal img_goal;
    img_goal.img_path = img_path;

    // 将goal发送给服务端，并设置回调函数
    client.sendGoal(img_goal, &doneCb, &activeCb, &feedbackCb);

    // 循环执行
    ros::spin();
    return 0;
}

File: scripts/img_client.py
# coding=utf-8
import rospy
import roslib
roslib.load_manifest('image_process')
import actionlib
import cv2
import numpy as np

from image_process.msg import ProcImgAction, ProcImgGoal


# 自定义函数，用于将字符串形式的数据转换成uint8数字表示的NDArray
# 在Python中Message的uint8数组以字符串形式传输的，这点需要注意
def cvtArray2Img(array, img_height, img_width):
    restore_img = np.zeros([img_height, img_width, 3], np.uint8)
    print "receiving and converting image data..."
    for i in range(img_height):
        for j in range(img_width):
            index = i*img_width+j
            # 利用ord()函数将单个字母转换成对应的字符编码
            b = ord(array[index])
            g = ord(array[index+img_height*img_width])
            r = ord(array[index+img_height*img_width*2])
            restore_img[i, j, :] = [b, g, r]
    return restore_img

# 当Action完成后的回调函数
def doneCb(state, result):
    # 新建一个Mat并将传输的字符串数据转换成OpenCV能识别的Mat类型
    processed_img = cvtArray2Img(
        result.processed_img, result.img_height, result.img_width)

    # 如果原图太大，将它缩小，这样利于展示
    if processed_img.shape[0] > 1200:
        ratio = 1200.0/result.img_width
        processed_img = cv2.resize(processed_img, None, None, ratio, ratio)

    rospy.loginfo("Finished processing image with %d x %d pixels...",
                  result.img_height, result.img_width)

    # 展示图片
    cv2.imshow("processed_img", processed_img)
    cv2.waitKey(0)

# 当Action启动后的回调函数
def activeCb():
    rospy.loginfo("Contected to server,start processing image...")

# 当收到Feedback后的回调函数
def feedbackCb(feedback):
    rospy.loginfo("percent complete:%.2f%%", feedback.complete_percent)


if __name__ == "__main__":
    # 用户输入图片路径
    img_path = raw_input("Input image path:\n")

    # 初始化并指定节点名称
    rospy.init_node("img_client")
    # 创建客户端并指定要连接到的服务端名称
    client = actionlib.SimpleActionClient("img_process", ProcImgAction)

    # 等待服务端
    rospy.loginfo("Waiting for action server to start.")
    client.wait_for_server()
    rospy.loginfo("Action server started, sending goal.")

    # 创建Action的goal，传入的是图片路径字符串
    img_goal = ProcImgGoal()
    img_goal.img_path = img_path

    # 将goal发送给服务端，并设置回调函数
    client.send_goal(img_goal, doneCb, activeCb, feedbackCb)

    # 循环执行
    rospy.spin()
